### PR TITLE
[Community Pipelines] Accelerate inference of AnimateDiff by IPEX on CPU

### PR DESCRIPTION
**Pull request by ustcuna**
Hi, this pipeline aims to speed up the inference of AnimateDiff on Intel Xeon CPUs on Linux. It is much alike the previous one I proposed and merged [#6683 ](https://github.com/huggingface/diffusers/pull/3105) for SDXL.

By using this optimized pipeline, we can get about 1.5-2.2 times performance acceleration with BFloat16 on fifth generation of Intel Xeon CPUs, code-named [[Emerald Rapids]](https://en.wikipedia.org/wiki/Emerald_Rapids).
It is also recommended to run on Pytorch/IPEX v2.0 and above to get the best performance boost.
The main profits are illustrated as below, which are the same with our previous PR:
  -  For Pytorch/IPEX v2.0 and above, it benefits from MHA optimization with Flash Attention and [TorchScript mode](https://intel.github.io/intel-extension-for-pytorch/latest/tutorials/examples.html#torchscript-mode) optimization in IPEX.
  -  For Pytorch/IPEX v1.13, it benefits from [TorchScript mode](https://intel.github.io/intel-extension-for-pytorch/latest/tutorials/examples.html#torchscript-mode) optimization in IPEX.

Below are the tables which show the test results for AnimateDiff-Lightning (a model that is a distilled version of AnimateDiff SD1.5 v2, a lightning-fast text-to-video generation model which uses AnimateDiff pipeline) with 1/2/4/8 steps on Intel® Xeon® Platinum 8582C Processor (60cores/socket, 1socket) w/ data type BF16:
image
Could u pls help to review? Thanks!

### CODE REVIEW

**Comment by ustcuna** [comment]
Hi @patrickvonplaten and @pcuenca, could you pls help to review this PR? since this one uses almost the same optimization methods and is with almost the same code format like the previous one I proposed and merged for SDXL pipeline[#6683 ](https://github.com/huggingface/diffusers/pull/6683) . Thanks a lot!

File: examples/community/README.md [hunk 1]
```diff
@@ -4123,6 +4124,117 @@ output_frames[0].save(output_video_path, save_all=True,
 
 ```
 
+### AnimateDiff on IPEX
+
+This diffusion pipeline aims to accelarate the inference of AnimateDiff on Intel Xeon CPUs with BF16/FP32 precision using [IPEX](https://github.com/intel/intel-extension-for-pytorch).
```
Comment (a-r-r-o-w) [inline examples/community/README.md:4129]:
```suggestion
This diffusion pipeline aims to accelerate the inference of AnimateDiff on Intel Xeon CPUs with BF16/FP32 precision using [IPEX](https://github.com/intel/intel-extension-for-pytorch).
```

File: examples/community/README.md [hunk 2]
```diff
@@ -4123,6 +4124,117 @@ output_frames[0].save(output_video_path, save_all=True,
...
+dtype = torch.float32
+
+prompt = "A girl smiling"
+step = 8  # Options: [1,2,4,8]
+repo = "ByteDance/AnimateDiff-Lightning"
+ckpt = f"animatediff_lightning_{step}step_diffusers.safetensors"
+base = "emilianJR/epiCRealism"  # Choose to your favorite base model.
+
+adapter = MotionAdapter().to(device, dtype)
+adapter.load_state_dict(load_file(hf_hub_download(repo ,ckpt), device=device))
```
Comment (a-r-r-o-w) [inline examples/community/README.md:4188]:
```suggestion
adapter.load_state_dict(load_file(hf_hub_download(repo, ckpt), device=device))
```

File: examples/community/README.md [hunk 3]
```diff
@@ -4123,6 +4124,117 @@ output_frames[0].save(output_video_path, save_all=True,
...
+device = "cpu"
+dtype = torch.float32
+
+prompt = "A girl smiling"
+step = 8  # Options: [1,2,4,8]
+repo = "ByteDance/AnimateDiff-Lightning"
+ckpt = f"animatediff_lightning_{step}step_diffusers.safetensors"
+base = "emilianJR/epiCRealism"  # Choose to your favorite base model.
+
+adapter = MotionAdapter().to(device, dtype)
```
Comment (a-r-r-o-w) [inline examples/community/README.md:4187]: Would actually prefer if `from_pretrained` could be used consistently but this is no problem either

File: examples/community/pipeline_animatediff_ipex.py [hunk 4]
```diff
@@ -0,0 +1,1039 @@
...
+        >>> with torch.cpu.amp.autocast(enabled=True, dtype=torch.bfloat16):
+        >>>     output = pipe(prompt = "A girl smiling", guidance_scale=1.0, num_inference_steps = step)
+
+        >>> frames = output.frames[0]
+        >>> export_to_gif(frames, "animation.gif")
+        ```
+"""
+
+
+class AnimateDiffPipelineIpex(AnimateDiffPipeline):
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:93]: Could you derive `DiffusionPipeline`, `StableDiffusionMixin`, `TextualInversionLoaderMixin`, `IPAdapterMixin`, `LoraLoaderMixin`, `FreeInitMixin` instead? We don't use our pipelines as parent classes

File: examples/community/pipeline_animatediff_ipex.py [hunk 5]
```diff
@@ -0,0 +1,1039 @@
...
+            PNDMScheduler,
+            LMSDiscreteScheduler,
+            EulerDiscreteScheduler,
+            EulerAncestralDiscreteScheduler,
+            DPMSolverMultistepScheduler,
+        ],
+        feature_extractor: CLIPImageProcessor = None,
+        image_encoder: CLIPVisionModelWithProjection = None,
+    ):
+        # super().__init__()
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:144]:
```suggestion
        super().__init__()
```

File: examples/community/pipeline_animatediff_ipex.py [hunk 6]
```diff
@@ -0,0 +1,1039 @@
...
+        negative_prompt_embeds: Optional[torch.Tensor] = None,
+        ip_adapter_image: Optional[PipelineImageInput] = None,
+        ip_adapter_image_embeds: Optional[List[torch.Tensor]] = None,
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+        cross_attention_kwargs: Optional[Dict[str, Any]] = None,
+        clip_skip: Optional[int] = None,
+        callback_on_step_end: Optional[Callable[[int, int, Dict], None]] = None,
+        callback_on_step_end_tensor_inputs: List[str] = ["latents"],
+        **kwargs,
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:596]:
```suggestion
```

File: examples/community/pipeline_animatediff_ipex.py [hunk 7]
```diff
@@ -0,0 +1,1039 @@
...
+                "callback",
+                "1.0.0",
+                "Passing `callback` as an input argument to `__call__` is deprecated, consider using `callback_on_step_end`",
+            )
+        if callback_steps is not None:
+            deprecate(
+                "callback_steps",
+                "1.0.0",
+                "Passing `callback_steps` as an input argument to `__call__` is deprecated, consider using `callback_on_step_end`",
+            )
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:687]: We don't need to add deprecated code to newer pipelines

```suggestion
```

File: examples/community/pipeline_animatediff_ipex.py [hunk 8]
```diff
@@ -0,0 +1,1039 @@
...
+        width = width or self.unet.config.sample_size * self.vae_scale_factor
+
+        num_videos_per_prompt = 1
+
+        # 1. Check inputs. Raise error if not correct
+        self.check_inputs(
+            prompt,
+            height,
+            width,
+            callback_steps,
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:700]:
```suggestion
```

File: examples/community/pipeline_animatediff_ipex.py [hunk 9]
```diff
@@ -0,0 +1,1039 @@
...
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.check_inputs
+    def check_inputs(
+        self,
+        prompt,
+        height,
+        width,
+        callback_steps,
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:458]:
```suggestion
```

File: examples/community/pipeline_animatediff_ipex.py [hunk 10]
```diff
@@ -0,0 +1,1039 @@
...
+        callback_on_step_end_tensor_inputs=None,
+    ):
+        if height % 8 != 0 or width % 8 != 0:
+            raise ValueError(f"`height` and `width` have to be divisible by 8 but are {height} and {width}.")
+
+        if callback_steps is not None and (not isinstance(callback_steps, int) or callback_steps <= 0):
+            raise ValueError(
+                f"`callback_steps` has to be a positive integer but is {callback_steps} of type"
+                f" {type(callback_steps)}."
+            )
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:473]:
```suggestion
```

File: examples/community/pipeline_animatediff_ipex.py [hunk 11]
```diff
@@ -0,0 +1,1039 @@
...
+
+                        latents = callback_outputs.pop("latents", latents)
+                        prompt_embeds = callback_outputs.pop("prompt_embeds", prompt_embeds)
+                        negative_prompt_embeds = callback_outputs.pop("negative_prompt_embeds", negative_prompt_embeds)
+
+                    # call the callback, if provided
+                    if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                        progress_bar.update()
+                        if callback is not None and i % callback_steps == 0:
+                            callback(i, t, latents)
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:830]:
```suggestion
```

File: examples/community/pipeline_animatediff_ipex.py [hunk 12]
```diff
@@ -0,0 +1,1039 @@
...
+        negative_prompt_embeds: Optional[torch.Tensor] = None,
+        ip_adapter_image: Optional[PipelineImageInput] = None,
+        ip_adapter_image_embeds: Optional[List[torch.Tensor]] = None,
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+        cross_attention_kwargs: Optional[Dict[str, Any]] = None,
+        clip_skip: Optional[int] = None,
+        callback_on_step_end: Optional[Callable[[int, int, Dict], None]] = None,
+        callback_on_step_end_tensor_inputs: List[str] = ["latents"],
+        **kwargs,
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:872]:
```suggestion
```

File: examples/community/pipeline_animatediff_ipex.py [hunk 13]
```diff
@@ -0,0 +1,1039 @@
...
+                "callback",
+                "1.0.0",
+                "Passing `callback` as an input argument to `__call__` is deprecated, consider using `callback_on_step_end`",
+            )
+        if callback_steps is not None:
+            deprecate(
+                "callback_steps",
+                "1.0.0",
+                "Passing `callback_steps` as an input argument to `__call__` is deprecated, consider using `callback_on_step_end`",
+            )
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:888]:
```suggestion
```

File: examples/community/pipeline_animatediff_ipex.py [hunk 14]
```diff
@@ -0,0 +1,1039 @@
...
+        width = width or self.unet.config.sample_size * self.vae_scale_factor
+
+        num_videos_per_prompt = 1
+
+        # 1. Check inputs. Raise error if not correct
+        self.check_inputs(
+            prompt,
+            height,
+            width,
+            callback_steps,
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_animatediff_ipex.py:901]:
```suggestion
```

**Review by a-r-r-o-w — COMMENTED** [review]
thank you for working on this and your patience! looks mostly good to me although I'm unable to test on a Xeon CPU at the moment. Have a few suggestions, could you please address them?

**Comment by ustcuna** [comment]
Hi @a-r-r-o-w , thank u so much for the detailed reviewing! So thoughtful and professional that helps improve the code with higher quality.🙂 With the new commit, I almost address all the suggestions you commented, could u pls help to review again? Except for the `adapter.from_pretrained` one I did not change, since here I was intending to totally keep the same example with the AnimateDiff-Lightning Diffusers Usage [ByteDance/AnimateDiff-Lightning](https://huggingface.co/ByteDance/AnimateDiff-Lightning#diffusers-usage). Do we need to change it here to keep consistency? If so, I will change it to `adapter.from_pretrained` with a new commit. Again, thanks sincerely!

**Comment by a-r-r-o-w** [comment]
Thank you applying the review comments and the kind words! The `adapter.from_pretrained` is a personal style choice and not a problem at all for community pipelines, and so we don't need to change it. The tests also have all passed, which is great and we can probably merge, but one last thing that I wanted to ask why are we not supporting `cross_attention_kwargs` and `added_cond_kwargs` here? Will that require preparing the IP Adapter models, for example, if loaded?

**Comment by ustcuna** [comment]
> Thank you applying the review comments and the kind words! The `adapter.from_pretrained` is a personal style choice and not a problem at all for community pipelines, and so we don't need to change it. The tests also have all passed, which is great and we can probably merge, but one last thing that I wanted to ask was why are we not supporting `cross_attention_kwargs` and `added_cond_kwargs` here? Will that require preparing the IP Adapter models, for example, if loaded?

Hi @a-r-r-o-w , the reason why we did not support `cross_attention_kwargs` and `added_cond_kwargs` here is that when leveraging torch.jit.trace to optimize unet, NoneType can not be inputs to traced functions. Even if we can feed fake inputs for `cross_attention_kwargs` and `added_cond_kwargs` to trace unet and succeed, we still need to add a fake input for `cross_attention_kwargs` and `added_cond_kwargs` inside __call__ function when they are default None, otherwise errors will occur due to incompatible function arguments. But I do think that would be confused for users to understand...
Thus we decide not to support `cross_attention_kwargs` and `added_cond_kwargs` here. Do you have any better suggestions to address this? Looking forward to your opinion! Thx!

**Comment by a-r-r-o-w** [comment]
I see, that is insightful. I do not have a deep understanding of torch.jit.trace (or torchscript in general) due to limited experience, but I've encountered NoneType issues in the past, so it is understandable why it might be hard to support those arguments. Optional type hinting could probably help (?) but it will require many diffusers core code changes I believe, which is out of scope, or fake inputs like you mention but it's okay to not do it here as this is an example pipeline.

Thanks for your contribution!

**Comment by ustcuna** [comment]
> I see, that is insightful. I do not have a deep understanding of torch.jit.trace (or torchscript in general) due to limited experience, but I've encountered NoneType issues in the past, so it is understandable why it might be hard to support those arguments. Optional type hinting could probably help (?) but it will require many diffusers core code changes I believe, which is out of scope, or fake inputs like you mention but it's okay to not do it here as this is an example pipeline.
> 
> Thanks for your contribution!

Thanks a lot for your understanding of this situation, which is quit a headache!😀 I would like to thank u again for reviewing and merging the code！

**Pull request merged**